db/crud.go

```diff
@@ -630,6 +630,9 @@ func (db *Database) updateDoc(docid string, allowImport bool, expiry uint32, cal
 			return
 		}
 
+		// Test for shadower echo here before UpstreamRev gets set to CurrentRev
+		shadowerEcho = doc.CurrentRev == doc.UpstreamRev
```

Comment:
If I'm reading this properly, doc.CurrentRev and doc.Upstream rev are the incoming values here (because the callback doesn't mutate doc, it just returns body).

What happens for a new SDK write that's being imported for the first time?  Would doc.CurrentRev and doc.UpstreamRev both be nil in that case (and so incorrectly be identified as an echo)?